jplag.frontend.rust/src/main/java/de/jplag/rust/grammar/RustLexerBase.java

```diff
@@ -0,0 +1,94 @@
+package de.jplag.rust.grammar;
+
+import org.antlr.v4.runtime.*;
+
+public abstract class RustLexerBase extends Lexer {
+    protected RustLexerBase(CharStream input) {
+        super(input);
+    }
+
+    Token lt1;
+    Token lt2;
+
+    @Override
+    public Token nextToken() {
+        Token next = super.nextToken();
+
+        if (next.getChannel() == Token.DEFAULT_CHANNEL) {
+            // Keep track of the last token on the default channel.
+            this.lt2 = this.lt1;
+            this.lt1 = next;
+        }
+
+        return next;
+    }
+
+    public boolean SOF() {
+        return _input.LA(-1) <= 0;
+    }
+
+    public boolean next(char expect) {
+        return _input.LA(1) == expect;
+    }
+
+    public boolean floatDotPossible() {
+        int next = _input.LA(1);
+        // only block . _ identifier after float
+        if (next == '.' || next == '_')
+            return false;
+        if (next == 'f') {
+            // 1.f32
+            if (_input.LA(2) == '3' && _input.LA(3) == '2')
+                return true;
+            // 1.f64
+            if (_input.LA(2) == '6' && _input.LA(3) == '4')
+                return true;
+            return false;
+        }
+        if (next >= 'a' && next <= 'z')
+            return false;
+        if (next >= 'A' && next <= 'Z')
+            return false;
+        return true;
+    }
```

Comment:
Is this class generated? I'm not sure about the naming of the methods here.

Reply:
It's not generated, but it's taken from the Antlr4 grammar library. I agree about the naming.

Reply:
I think we shold refactor it, as it really is in bad condition.

Reply:
@robinmaisch, I think this is the only open issue here before we can merge.

Reply:
I think we are also waiting for the new `TokenPrinter` such that the stored annotated files can look good.

Reply:
Good call, I missed that.

Reply:
Now it should look a bit better. I cannot quite grasp what floatLiteralPossible and floatDotPossible really do, so I just tried to make them more readable without altering their behaviour.

Reply:
I think this is the best way to treat this code. From glancing at it, I am not sure either what it does (probably something regarding parsing [differently encoded float values](https://doc.rust-lang.org/std/primitive.f64.html)), but that just shows us how important well-written and well-documented code is.